### PR TITLE
:tada: backport-migrate CLI for creating steps from backported datasets

### PR DESCRIPTION
**Pull request by Marigold**
New tool for migrating datasets from MySQL into ETL. It automatically creates:

- ingest script
- garden step with YAML metadata
- grapher step

Running ingest script & etl steps then recreates the original dataset with proper data & metadata.

This is pretty similar to backports, but unlike them, it generates all files in ETL and is meant to be maintained. Backports only create snapshots in S3 and then dynamically populate ETL catalog (which is something we want to move away from).

It was motivated by COVID datasets migration to ETL. I was too lazy to use walkthrough and copy&paste metadata manually. This will save me tons of time and prevent pasting errors.

Migration leverages Snapshot of backported dataset (we could also load directly from MySQL, but we couldn't reuse existing backports code), it loads it in ingest script and creates new Snapshot that becomes basis for steps.

`backport-migrate` could be useful in these situations:
- Getting head-start when migrating datasets from MySQL to ETL
- Replacement for `backport://...` datasets that are not very transparent right now
- Enable us to remove `Import CSV` (which were used to fix data errors in existing datasets)

## How to review

Try running
```bash
poetry install
backport-migrate --dataset-id 5205 --namespace covid --short-name hospital__and__icu --no-backport
```

check generated outputs and then run
```
python snapshots/covid/latest/hospital__and__icu.py
etl covid/latest/hospital__and__icu --grapher
```

I started matching [snapshots without extension](https://github.com/owid/etl/pull/1132/files#diff-e6bc1da6afdc039bc1dc77bad1c6069064eef701aad2c6d49606b1a4bc3f39a4R277) and also added two helper methods `load_snapshot_dependency` and `load_dataset_dependency` (to avoid `cast`ing outputs). All the other changes are non essential for production and don't have to be thoroughly reviewed.

### CODE REVIEW

File: etl/helpers.py [hunk 1]
```diff
@@ -406,6 +409,18 @@ def load_dependency(
 
         return dataset
 
+    def load_snapshot_dependency(self) -> Snapshot:
```
Comment (pabloarosado) [inline etl/helpers.py:412]: Nice! You could change the walkthrough templates to use these functions instead of the generic `load_dependency`.

**Review by pabloarosado — COMMENTED** [review]
Good idea! I tried to run it and got:
```
(etl-py3.11) ➜  etl git:(backport-migrate) ✗ backport-migrate --dataset-id 5205 --namespace covid --short-name hospital__and__icu --no-backport

2023-05-23 09:57:19,564 - [bugsnag] WARNING - No API key configured, couldn't notify
Traceback (most recent call last):
  File "/Users/prosado/Documents/owid/repos/etl/.venv/bin/backport-migrate", line 6, in <module>
    sys.exit(cli())
             ^^^^^
  File "/Users/prosado/Documents/owid/repos/etl/.venv/lib/python3.11/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/prosado/Documents/owid/repos/etl/.venv/lib/python3.11/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/Users/prosado/Documents/owid/repos/etl/.venv/lib/python3.11/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/prosado/Documents/owid/repos/etl/.venv/lib/python3.11/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/prosado/Documents/owid/repos/etl/backport/migrate/migrate.py", line 80, in cli
    return migrate(
           ^^^^^^^^
  File "/Users/prosado/Documents/owid/repos/etl/backport/migrate/migrate.py", line 147, in migrate
    _generate_metadata_yaml(namespace, version, short_name, pb.short_name)
  File "/Users/prosado/Documents/owid/repos/etl/backport/migrate/migrate.py", line 189, in _generate_metadata_yaml
    ds = create_dataset("/tmp/migrate", backport_short_name, new_short_name=short_name)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/prosado/Documents/owid/repos/etl/etl/backport_helpers.py", line 114, in create_dataset
    config = load_config(short_name)
             ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/prosado/Documents/owid/repos/etl/etl/backport_helpers.py", line 49, in load_config
    return GrapherConfig.parse_file(snap.path)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/main.py", line 561, in pydantic.main.BaseModel.parse_file
  File "pydantic/parse.py", line 57, in pydantic.parse.load_file
  File "/opt/homebrew/Cellar/python@3.11/3.11.2_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/pathlib.py", line 1050, in read_bytes
    with self.open(mode='rb') as f:
         ^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.2_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/pathlib.py", line 1044, in open
    return io.open(self, mode, buffering, encoding, errors, newline)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/Users/prosado/Documents/owid/repos/etl/data/snapshots/backport/latest/dataset_5205_covid_2019__hospital__and__icu_config.json'
```

I haven't looked into it, but I run `etl --backport dataset_5205_covid_2019__hospital__and__icu_values` and worked well.

File: etl/helpers.py (hunk 1 above)
Comment (Marigold) [reply etl/helpers.py:412]: Oh right, gonna refactor it in the future PR.

**Comment by Marigold** [comment]
> I haven't looked into it, but I run etl --backport dataset_5205_covid_2019__hospital__and__icu_values and worked well.

That's a bug, will fix it in future PR. Thanks for review!

**Pull request merged**